doc/manual/src/architecture/file-system-object.md

```diff
@@ -0,0 +1,63 @@
+# File System Object
+
+Nix uses a simplified model of the file system, which consists of file system objects.
+Every file system object is one of the following:
+
+ - File
+
+   - A possibly empty sequence of bytes for contents
+   - A single boolean representing the [executable](https://en.m.wikipedia.org/wiki/File-system_permissions#Permissions) permission
+
+ - Directory
+
+   Mapping of names to child file system objects
+
+ - [Symbolic link](https://en.m.wikipedia.org/wiki/Symbolic_link)
+
+   A file system path that may point anywhere
```

Comment:
We didn't define a file system path, but let's not open that can of worms. Not that it shouldn't be opened, but it doesn't seem productive. It will be more _interesting_ to define it when windows support is a thing, fwiw.

Reply:
Yeah I was thinking about that. Get ready for https://simonsapin.github.io/wtf-8/ !

Similarly in the code we should probably make name types which enforce things separate from `StorePath`, `OutputsSpec`, `DerivedPath`, etc.